[cxx-interop] Add support for operator! overloading

bulbazord opened the pull request:
This patch enables operator overloading for operator! when cxx interop is enabled. This is my first patch related to cxx interop, let me know if I should be doing something differently.

@plotfi @zoecarver @hyp

bulbazord committed:
[cxx-interop] Add support for operator! overloading

plotfi commented:
Nice work Alex. :-)

bulbazord commented:
@swift-ci please test

zoecarver commented on test/Interop/Cxx/operators/Inputs/member-inline.h:
```diff
@@ -18,6 +18,17 @@ struct LoadableIntWrapper {
+    return !value;
+  }
+
+  bool operator()() {
+    return value;
+  }
```
Why did this get added?

bulbazord replied:
Ah, this was a leftover from when I was working on the test initially. I'll remove it as it's unused.

zoecarver reviewed (approved):
Seems good. Thanks.

bulbazord commented:
Okay, I updated the PR. Here's the changelog:
- Removed unneeded operator overload from `LoadableBoolWrapper`
- Modified the signature of `LoadableBoolWrapper::operator!` to return a `LoadableBoolWrapper`. This makes it more consistent with the other tests.
- Added tests to `member-inline-typechecker.swift` and `member-inline-module-interface.swift` as requested. I also updated `member-inline-silgen.test` to keep it consistent.

Let me know if there's anything else you'd like to see changed.

zoecarver commented on test/Interop/Cxx/operators/member-inline-silgen.swift:
```diff
@@ -13,6 +13,17 @@ public func sub(_ lhs: inout LoadableIntWrapper, _ rhs: LoadableIntWrapper) -> L
+
+// CHECK: bb0([[SELF:%.*]] : $*LoadableBoolWrapper):
+// CHECK:   [[SELFACCESS:%.*]] = begin_access [modify] [static] [[SELF]] : $*LoadableBoolWrapper
+// CHECK:   [[OP:%.*]] = function_ref @_ZN19LoadableBoolWrapperntEv : $@convention(c) (@inout LoadableBoolWrapper) -> LoadableBoolWrapper
```
The mangling on windows is different :/

zoecarver commented:
This looks great. Thanks again!

bulbazord commented:
Yes, if you could please trigger the bots to show me the Windows mangling that would be great! I don't have access to a windows machine currently.

bulbazord commented:
@swift-ci please test

bulbazord commented:
@zoecarver Could you trigger the test bots so I can get an idea of what the windows manglings will be?

drodriguez commented:
@swift-ci please smoke test

bulbazord commented:
Ok, got the test fixed up for windows. @drodriguez Can you trigger the tests again for me please? Thanks!

drodriguez commented:
@swift-ci please smoke test

bulbazord commented:
@zoecarver Think this is good to go in now? Or should we wait for the Linux Platform to be unborked?

drodriguez commented:
@swift-ci please test Linux platform
